rfcs/proposed/parallel_block_for_task_arena/README.md

```diff
@@ -0,0 +1,189 @@
+  will be released from arena?
+  * What if we indicated that threads should leave arena after the "Parallel Block"?
+  * What if we just indicated the end of the "Parallel Block"?
+
+The extended state machine aims to answer these questions.
+* The first call to the “Start of PB” will transition into the “Parallel Block” state.
+* The last call to the “End of PB” will transition back to the “Delayed leave” state
+  or into the "One-time Fast leave" if it is indicated that threads should leave sooner.
+* Concurrent or nested calls to the “Start of PB” or the “End of PB”
+  increment/decrement reference counter.
+
+<img src="parallel_block_state_final.png" width=800>
+
+Let's consider the semantics that an API for explicit parallel blocks can provide:
+* Start of a parallel block:
+  * Indicates the point from which the scheduler can use a hint and keep threads in the arena
+    for longer.
+  * Serves as a warm-up hint to the scheduler:
+    * Allows worker threads to be available by the time real computation starts.
+* "Parallel block" itself:
+  * Scheduler can implement different policies to retain threads in the arena.
+  * The semantics for retaining threads is a hint to the scheduler;
+    thus, no real guarantee is provided. The scheduler can ignore the hint and
+    move threads to another arena or to sleep if conditions are met.
+* End of a parallel block:
+  * Indicates the point from which the scheduler may drop a hint and
+    no longer retain threads in the arena.
+  * Indicates that arena should enter the “One-time Fast leave” thus workers can leave sooner.
+    * If work was submitted immediately after the end of the parallel block,
+      the default arena "workers leave" state will be restored.
+    * If the default "workers leave" state was the "Fast leave" the result is NOP.
+
+
+### Proposed API
+
+```cpp
+class task_arena {
+    enum class workers_leave : /* unspecified type */ {
+        fast = /* unspecifed */,
+        delayed = /* unspecifed */
+    };
+
+    task_arena(int max_concurrency = automatic, unsigned reserved_for_masters = 1,
+               priority a_priority = priority::normal,
+               workers_leave a_workers_leave = workers_leave::delayed);
+
+    task_arena(const constraints& constraints_, unsigned reserved_for_masters = 1,
+               priority a_priority = priority::normal,
+               workers_leave a_workers_leave = workers_leave::delayed);
+
+    void initialize(int max_concurrency, unsigned reserved_for_masters = 1,
+                    priority a_priority = priority::normal,
+                    workers_leave a_workers_leave = workers_leave::delayed);
+
+    void initialize(constraints a_constraints, unsigned reserved_for_masters = 1,
+                    priority a_priority = priority::normal,
+                    workers_leave a_workers_leave = workers_leave::delayed);
+
+    void start_parallel_block();
+    void end_parallel_block(bool set_one_time_fast_leave = false);
```

Comment:
More on the API names.
- There might be alternative names to consider instead of "block" - e.g., "stage", "phase", "region". Personally, I like any of these three, and especially "phase", quite more than "block".
- `set_one_time_fast_leave` is too verbose and, due to "set", suggests that it might have impact beyond the call. Something like `with_fast_leave` could be better.

Reply:
Seems like "region" might be confused (or wrongly associated) with OpenMP's "parallel Region" term (we have different guarantees). Therefore, I think "stage" or "phase" are more preferred names.